// fa/index.js
// Part of <https://miracle.systems/p/walkner-fa> licensed under <CC BY-NC-SA 4.0>

'use strict';

const generateNextNo = require('./generateNextNo');
const canAccess = require('./canAccess');

exports.DEFAULT_CONFIG = {
  mongooseId: 'mongoose',
  expressId: 'express',
  userId: 'user',
  settingsId: 'settings',
  mailSenderId: 'mail/sender',
  uploadsDest: './data/uploads/fa',
  maxAttachmentSize: 10 * 1024 * 1024,
  maxFileSize: 10 * 1024 * 1024,
  tplFillerExe: 'FaTplFiller.exe'
};

exports.models = [
  require('./models/faReqTpl'),
  require('./models/faAssetGroup'),
  require('./models/faAssetClass'),
  require('./models/faDepKey'),
  require('./models/faCostCenter'),
  require('./models/faOt'),
  require('./models/faLt'),
  require('user/models/user')
];

exports.republishTopics = [
  'fa.assetGroups.**',
  'fa.assetClasses.**',
  'fa.depKeys.**',
  'fa.costCenters.**',
  'fa.ot.**',
  'fa.lt.**'
];

exports.recordTopics = {
  warning: [
    'fa.assetGroups.deleted',
    'fa.assetClasses.deleted',
    'fa.depKeys.deleted',
    'fa.costCenters.deleted',
    'fa.ot.deleted',
    'fa.lt.deleted'
  ]
};

exports.userPrivileges = [
  'FA:VIEW',
  'FA:VIEW:ALL',
  'FA:MANAGE',
  'FA:MANAGE:FILES',
  'FA:LT:committee',
  'FA:OT:VIEW',
  'FA:OT:MANAGE',
  'FA:OT:ADD',
  'FA:OT:verify',
  'FA:OT:record',
  'FA:LT:VIEW',
  'FA:LT:MANAGE',
  'FA:LT:ADD',
  'FA:LT:verify',
  'FA:LT:verifyDocument',
  'FA:LT:acceptFinance',
  'FA:LT:acceptDepartment',
  'FA:LT:record'
];

exports.requiredModules = 'mongoose';

exports.optionalModules = {
  'express user settings': require('./routes'),
  'mailSender': [
    require('./notifier'),
    require('./reminder')
  ]
};

exports.start = (app, module) =>
{
  module.DICTIONARIES = {
    assetGroups: 'FaAssetGroup',
    assetClasses: 'FaAssetClass',
    depKeys: 'FaDepKey',
    costCenters: 'FaCostCenter'
  };

  module.tmpAttachments = new Map();
  module.generateNextNo = generateNextNo.bind(null, app, module);
  module.canAccess = canAccess.bind(null, app, module);
};
