// sio/index.js
// Part of <https://miracle.systems/p/walkner-wmes> licensed under <CC BY-NC-SA 4.0>

'use strict';

const {exec} = require('child_process');
const _ = require('lodash');
const engineIo = require('engine.io');
const socketIo = require('socket.io');
const SocketIoMultiServer = require('./SocketIoMultiServer');
const setUpRoutes = require('./routes');
let pmx = null;

try { pmx = require('pmx'); }
catch (err) {} // eslint-disable-line no-empty

engineIo.Server.errors.SERVER_UNAVAILABLE = 503;
engineIo.Server.errorMessages[503] = 'Server unavailable';

exports.DEFAULT_CONFIG = {
  httpServerIds: ['httpServer'],
  expressId: 'express',
  userId: 'user',
  pingsId: 'pings',
  path: '/sio',
  socketIo: {
    pingInterval: 30000,
    pingTimeout: 10000
  },
  netstatCmd: ''
};

exports.start = (app, module, done) =>
{
  module.config.socketIo = Object.assign({}, module.config.socketIo, {
    path: module.config.path,
    transports: ['websocket', 'xhr-polling'],
    serveClient: false,
    allowRequest: checkRequest
  });

  let socketCount = 0;
  const probes = {
    currentUsersCounter: null,
    totalConnectionTime: null,
    totalConnectionCount: null
  };

  if (pmx)
  {
    const pmxProbe = pmx.probe();

    probes.currentUsersCounter = pmxProbe.counter({name: 'sio:currentUsers'});
    probes.totalConnectionTime = pmxProbe.histogram({name: 'sio:totalConnectionTime', measurement: 'sum'});
    probes.totalConnectionCount = pmxProbe.histogram({name: 'sio:totalConnectionCount', measurement: 'sum'});
  }

  const multiServer = new SocketIoMultiServer();

  app.onModuleReady(module.config.httpServerIds, () =>
  {
    _.forEach(module.config.httpServerIds, (httpServerId) =>
    {
      multiServer.addServer(app[httpServerId].server);
    });

    startSocketIo();

    done();
  });

  function checkRequest(req, done)
  {
    const allServersAvailable = module.config.httpServerIds.every(httpServerId => app[httpServerId].isAvailable());

    if (!allServersAvailable)
    {
      return done(engineIo.Server.errors.SERVER_UNAVAILABLE, false);
    }

    if (!req.headers.origin)
    {
      return done(null, true);
    }

    return module.checkRequest(req, done);
  }

  function startSocketIo()
  {
    const sio = socketIo(multiServer, module.config.socketIo);

    module = app[module.name] = Object.assign(sio, module);

    app.onModuleReady(
      [
        module.config.userId,
        module.config.expressId
      ],
      setUpRoutes.bind(null, app, module)
    );

    sio.sockets.setMaxListeners(25);

    module.on('connection', socket =>
    {
      ++socketCount;

      socket.handshake.connectedAt = socket.handshake.issued;
      socket.handshake.binary = socket.handshake.query.binary === '1';

      if (!socket.handshake.binary)
      {
        module.warn('Socket does not support binary messages.', {
          socket: socket.handshake
        });
      }

      if (pmx)
      {
        probes.currentUsersCounter.inc();
      }

      if (app[module.config.pingsId])
      {
        app[module.config.pingsId].recordHttpRequest(socket.conn.request, socket.handshake.user);
      }

      app.broker.publish('sockets.connected', {
        socket,
        socketCount
      });

      socket.on('disconnect', () =>
      {
        --socketCount;

        if (pmx)
        {
          probes.totalConnectionCount.update(1);
          probes.totalConnectionTime.update((Date.now() - socket.handshake.connectedAt) / 1000);
          probes.currentUsersCounter.dec();
        }

        app.broker.publish(`sockets.disconnected.${socket.id}`, {
          socket,
          socketCount
        });
      });

      socket.on('time', reply =>
      {
        if (typeof reply === 'function')
        {
          reply(Date.now(), app.options.serverTimezone || 'Europe/Warsaw');
        }
      });

      socket.on('time2', reply =>
      {
        if (typeof reply === 'function')
        {
          reply(null, {
            ts: Date.now(),
            tz: app.options.serverTimezone || 'Europe/Warsaw'
          });
        }
      });
    });
  }
};
