// fa/routes/lt/add.js
// Part of <https://miracle.systems/p/walkner-fa> licensed under <CC BY-NC-SA 4.0>

'use strict';

const step = require('h5.step');
const moment = require('moment');

module.exports = (app, module, req, res, next) =>
{
  const {
    user,
    FaLt,
    generateNextNo
  } = module;

  step(
    function()
    {
      if (req.body.kind !== 'merge')
      {
        ['mergeInventoryNo', 'mergeLineSymbol', 'mergeType', 'mergeNotes'].forEach(prop =>
        {
          req.body[prop] = '';
        });
      }

      this.doc = new FaLt({
        createdAt: new Date(),
        createdBy: user.createUserInfo(req.session.user, req),
        updatedAt: null,
        updatedBy: null,
        stage: 'protocol',
        stageChangedAt: {
          protocol: null,
          verify: null,
          acceptOwner: null,
          acceptDepartment: null,
          acceptFinance: null,
          record: null
        },
        stageChangedBy: {
          protocol: null,
          verify: null,
          acceptOwner: null,
          acceptDepartment: null,
          acceptFinance: null,
          record: null
        },
        kind: req.body.kind,
        no: null,
        date: moment.utc(moment().format('YYYY-MM-DD'), 'YYYY-MM-DD').toDate(),
        inventoryNo: '',
        assetName: '',
        costCenter: null,
        applicant: null,
        cause: '',
        initialValue: 0,
        deprecationValue: 0,
        netValue: 0,
        economicDeprecationValue: 0,
        economicNetValue: 0,
        verifyNotes: '',
        mergeInventoryNo: req.body.mergeInventoryNo || '',
        mergeLineSymbol: req.body.mergeLineSymbol || '',
        mergeType: req.body.mergeType || null,
        mergeNotes: req.body.mergeNotes || '',
        buyerName: '',
        buyerAddress: '',
        saleValue: 0,
        otherNotes: req.body.kind === 'other' ? req.body.otherNotes : '',
        sapNo: '',
        accountingNo: '',
        changes: []
      });

      generateNextNo(this.doc, this.next());
    },
    function(err, unlockNoGenerator)
    {
      if (err)
      {
        return this.skip(app.createError(
          `Failed to generate the next document number: ${err.message}`,
          'NO_GENERATION_FAILURE',
          500
        ));
      }

      this.unlockNoGenerator = unlockNoGenerator;

      this.doc.save(this.next());
    },
    function(err)
    {
      if (err)
      {
        return this.skip(app.createError(
          `Failed to save a new document: ${err.message}`,
          'SAVE_FAILURE',
          500
        ));
      }
    },
    function(err)
    {
      if (this.unlockNoGenerator)
      {
        this.unlockNoGenerator();
      }

      if (err)
      {
        return next(err);
      }

      res.json(this.doc);

      app.broker.publish(`${FaLt.TOPIC_PREFIX}.added`, {
        model: this.doc,
        user: this.doc.createdBy
      });
    }
  );
};
